pkg/daemon/bgp/bgp.go

```diff
@@ -0,0 +1,353 @@
+			manager.routerV6Address = existLinkAddress[0].IP
+		} else {
+			manager.routerV4Address = existLinkAddress[0].IP
+		}
+	case 2:
+		if (existLinkAddress[0].IP.To4() != nil && existLinkAddress[1].IP.To4() == nil) ||
+			(existLinkAddress[0].IP.To4() == nil && existLinkAddress[1].IP.To4() != nil) {
+			for _, addr := range existLinkAddress {
+				if addr.IP.To4() == nil {
+					manager.routerV6Address = addr.IP
+				}
+				manager.routerV4Address = addr.IP
+			}
+
+			// Use v4 address as routerID by default if v4/v6 addresses exist at the same time.
+			manager.routerID = manager.routerV4Address.String()
+			break
+		}
+		fallthrough
+	default:
+		defaultV4Route, err := containernetwork.GetDefaultRoute(netlink.FAMILY_V4)
+		if err != nil && err != daemonutils.NotExist {
+			return nil, fmt.Errorf("failed to get v4 default route: %v", err)
+		}
+
+		defaultV6Route, err := containernetwork.GetDefaultRoute(netlink.FAMILY_V6)
+		if err != nil && err != daemonutils.NotExist {
+			return nil, fmt.Errorf("failed to get v6 default route: %v", err)
+		}
+
+		for _, addr := range existLinkAddress {
+			if defaultV4Route != nil {
+				if addr.IP.Equal(defaultV4Route.Src) || addr.IPNet.Contains(defaultV4Route.Gw) {
+					manager.routerV4Address = addr.IP
+				}
+			}
+
+			if defaultV6Route != nil {
+				if addr.IP.Equal(defaultV6Route.Src) || addr.IPNet.Contains(defaultV6Route.Gw) {
+					manager.routerV6Address = addr.IP
+				}
+			}
+		}
+
+		if manager.routerV4Address == nil && manager.routerV6Address == nil {
+			return nil, fmt.Errorf("failed to find valid address for bgp router")
+		}
+
+		if manager.routerV4Address != nil {
+			// Use v4 address as routerID by default if v4/v6 addresses exist at the same time.
+			manager.routerID = manager.routerV4Address.String()
+		} else {
+			manager.routerID = manager.routerV6Address.String()
+		}
+	}
+
+	go manager.bgpServer.Serve()
+	return manager, nil
+}
+
+func (m *Manager) RecordPeer(address, password string, asn int, gracefulRestartTime int32) {
+	if gracefulRestartTime == 0 {
+		gracefulRestartTime = 300
+	}
+
+	m.peerMap[address] = &peerInfo{
+		address:                address,
+		asn:                    asn,
+		gracefulRestartSeconds: uint32(gracefulRestartTime),
+		password:               password,
+	}
+}
+
+func (m *Manager) RecordSubnet(cidr *net.IPNet) {
+	m.subnetMap[cidr.String()] = cidr
+}
+
+func (m *Manager) RecordIPInstance(ip net.IP) {
+	m.ipInstanceMap[ip.String()] = ip
+}
+
+func (m *Manager) ResetInfos() {
+	m.peerMap = map[string]*peerInfo{}
+	m.subnetMap = map[string]*net.IPNet{}
+	m.ipInstanceMap = map[string]net.IP{}
+}
+
+func (m *Manager) TryStart(asn uint32) error {
+	if m.localASN == 0 {
+		m.localASN = asn
+	} else if m.localASN != asn {
+		return fmt.Errorf("can not restart bgp manager (local AS number: %v) with a different AS number %v",
+			m.localASN, asn)
+	} else {
+		return nil
+	}
+
+	return m.bgpServer.StartBgp(context.Background(), &api.StartBgpRequest{
```

Comment:
these lines will be never touched

Reply:
Yes it will, if `m.localASN` is initialized, it will be touched.